[FLINK-10868][flink-yarn] Enforce maximum TMs failure rate in ResourceManagers

HuangZhenQiu opened the pull request:
## What is the purpose of the change

The pull request enforces maximum failed TMs in YarnResourceManager to achieve feature parity with legacy YarnFlinkResouceManager. This fix helps to make yarn cluster gracefully shutdown even when TM keeps on failure due to external resource caused issues, such as checkpoint files larget than hdfs folder quota.

## Brief change log
  - Add maximum failed TM threshold in ResourceManager 
  - When failure TM hit the threshold, cancel all of the pending slot requests in slot manager
## Verifying this change

This change added tests and can be verified as follows:
   - *add testOnContainersAllocatedWithFailure test case in YarnResourceManagerTest*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: Yarn (yes)
  - The S3 file system connector: (no)

## Documentation
  - Does this pull request introduce a new feature? (no)

suez1224 commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java:
```diff
@@ -145,6 +146,9 @@
 	private ConcurrentMap<String, InfoMessageListenerRpcGateway> infoMessageListeners;
 
+	/** The number of failed containers since the master became active. */
+	protected AtomicInteger failedContainerSoFar = new AtomicInteger(0);
```
Do we Atomic here? Is there concurrent access?

HuangZhenQiu replied:
There are three code paths that will create failed container. One is the container can't be started from YARN, which is notified by NMClient callback, the other two are failed to registerTastManager and disconnectTaskManager (When tasks fail). Thus, the concurrent is between akka thread and main thread?

suez1224 commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java:
```diff
@@ -381,11 +393,20 @@ public void onContainersAllocated(List<Container> containers) {
-						// and ask for a new one
-						requestYarnContainerIfRequired(container.getResource(), container.getPriority());
+
+						if (failedContainerSoFar.intValue() < maxFailedContainers) {
```
I think we need to fail the cluster when failedContainerSoFar exceed the threshold after line 362 as well, right?

HuangZhenQiu replied:
Yes, you are right. Will change also.

HuangZhenQiu commented:
@suez1224 Thanks for the input. I will revise the PR tonight.

walterddr commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java:
```diff
@@ -145,6 +147,12 @@
+	protected AtomicInteger failedContainerSoFar = new AtomicInteger(0);
+
+	/** Number of failed TaskManager containers before stopping the application. Default is  Integer.MAX_VALUE */
+	protected int maxFailedContainers = Integer.MAX_VALUE;
```
this should be `maximumAllowedTaskManagerFailureCount`, to avoid confusion since this is only used during `closeTaskManagerConnection` but not `closeJobManagerConnection`

HuangZhenQiu replied:
Agree.

walterddr commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/exceptions/MaximumFailedContainersException.java:
```diff
@@ -0,0 +1,35 @@
+/**
+ * Exception for {@link ResourceManager} when it identified that the maximum number of failed containers is hit.
+ */
+public class MaximumFailedContainersException extends ResourceManagerException {
```
probably rename to `MaximumFailedTaskManagerExceedingException` ?

HuangZhenQiu replied:
Good suggestion.

walterddr commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java:
```diff
@@ -825,6 +837,12 @@ protected void closeTaskManagerConnection(final ResourceID resourceID, final Exc
 			workerRegistration.getTaskExecutorGateway().disconnectResourceManager(cause);
+			failedContainerSoFar.getAndAdd(1);
+			if (failedContainerSoFar.intValue() >= maxFailedContainers) {
+				rejectAllPendingSlotRequests(new MaximumFailedContainersException(
```
do we need to do 
`slotManager.unregisterTaskManager` here to clear all the registered TaskManagers as well?

HuangZhenQiu replied:
No need. If other TMs can be still alive for restart strategy to make decision about whether to fail whole job. For per job cluster, if the job failed, the cluster will terminate by itself.

walterddr commented on flink-yarn/src/test/java/org/apache/flink/yarn/YarnResourceManagerTest.java:
```diff
@@ -512,4 +517,53 @@ public void testOnContainerCompleted() throws Exception {
+				// request second slot
+				registerSlotRequestFuture = resourceManager.runInMainThread(() -> {
+					rmServices.slotManager.registerSlotRequest(
+						new SlotRequest(new JobID(), new AllocationID(), resourceProfile1, taskHost));
```
can probably mock this so that no need to spy on the `slotManager` ? verify `request.completeExceptionally()` has been call should be suffice.

HuangZhenQiu replied:
Good suggestion. But it is SlotPool's PendingRequest will completeExceptionally, but this slot request instance right?

tillrohrmann commented:
Current test failure is unrelated.

HuangZhenQiu commented:
@tillrohrmann 
Thanks for you comments. 
1) Totally agreement. This PR should also support MesosResourceManager. Let me rephrase the title of the JIRA ticket.

2)Yes, It should be disabled by default. I can simply initialize the default value of maximum Number of allowed failure executor as Integer.MAX_VALUE.

3) There are several type of failure scenarios. 
- When start a new container, there is a namenode failover or hdfs is down. Then, the container can't fetch job jar to bootstrap.
- We allocate hdfs quota for each job's checkpoint folder. When quota is hit, containers will consistently fail, and restart strategy will restart the job by allocating more containers. In this condition, a job will continue to run for a while, but the job is actually in wrong state.

I prefer to add the threshold on job rather than whole cluster. But when i jump into implementation. I found it is hard to distinguish which allocated container is for which job master in YarnResourceManager. Any suggestion for it?

4) As you suggested in the initial conversation in Jira ticket, currently the MaximumFailedTaskManagerExceedingException is thrown to ExecutionGraph, then rely on different type of RestartStrategy to take action. As preventing infinite restart from restart strategy is definitely what we want, If so I will let MaximumFailedTaskManagerExceedingException extends from SuppressRestartsException. 

I will update the PR accordingly.

tillrohrmann commented:
2. What about introducing a failure rate instead of a total number of failures? We could say allow so and so many container failures per hour. If this number is exceeded, then fail all pending slot requests with a container failure rate exception.
3. By having a failure rate it might be ok to not account the failure for each job individually because the cluster can still recover after some time and won't be useless for future job submissions. Atm there is no way to find out the mapping between containers and jobs. The idea is that free TaskManager from another job can be used for serving slot requests of another job. It would actually be good if only unexpected container completions (e.g. if a container dies) would be counted for the failure rate. If a TM disconnects, then we should not increase the failure counter because it might still be there.
4. Yes temporarily we should let `MaximumFailedTaskManagerExceedingException` extend from `SuppressRestartException`. I think what we should do as a follow up is to extend the `RestartStrategy` such that it knows about the failure reason. That way, it would be possible to implement a custom `RestartStrategy` which can treat failures differently (e.g. `MaximumFailedTaskManagerExceedingException`).

HuangZhenQiu commented on flink-yarn/src/test/java/org/apache/flink/yarn/YarnResourceManagerTest.java:
```diff
@@ -521,4 +526,57 @@ public void testOnContainerCompleted() throws Exception {
+				when(mockNMClient.startContainer(eq(failedContainer), any())).thenThrow(new YarnException("Failed"));
+				resourceManager.onContainersAllocated(ImmutableList.of(failedContainer));
+				verify(rmServices.slotManager, times(1))
+					.rejectAllPendingSlotRequests(any(MaximumFailedTaskManagerExceedingException.class));
```
Agree. Changed accordingly.

HuangZhenQiu commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java:
```diff
@@ -301,12 +301,25 @@ public boolean registerSlotRequest(SlotRequest slotRequest) throws SlotManagerEx
+		}
+
+		pendingSlotRequests.clear();
+	}
```
It can be checked by verifying the pendingSlotRequests. Added in test case.

HuangZhenQiu commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java:
```diff
@@ -825,6 +843,12 @@ protected void closeTaskManagerConnection(final ResourceID resourceID, final Exc
+				rejectAllPendingSlotRequests(new MaximumFailedTaskManagerExceedingException(
+					String.format("Maximum number of failed container %d "
+						+ "is detected in Resource Manager", failedContainerSoFar.intValue())));
+			}
```
Agree. Removed.

HuangZhenQiu commented:
1) Added the same feature for MesosResourceManager.
2) Agree. Added a failure rate mechanism in resource manager.  It is similar with the FailureRateRestartStrategy.
3) Agree, removed the counter for disconnectTaskManager function.
4) Done.

suez1224 commented on flink-yarn/src/main/java/org/apache/flink/yarn/configuration/YarnConfigOptions.java:
```diff
@@ -83,8 +83,27 @@
+	public static final ConfigOption<Integer> CONTAINERS_FAILURE_RATE_INTERVAL =
+		key("yarn.containers-failure-rate-interval")
+		.defaultValue(300)
+		.withDeprecatedKeys("The interval for measuring failure rate of containers");
```
Should be withDescription here, not withDeprecatedKeys.

HuangZhenQiu replied:
Good catch. Updated.

tillrohrmann commented on docs/_includes/generated/mesos_configuration.html:
```diff
@@ -27,6 +27,11 @@
+        <tr>
+            <td><h5>mesos.maximum-failed-workers-per-interval</h5></td>
+            <td style="word-wrap: break-word;">-1</td>
+            <td>Maximum number of workers the system is going to reallocate in case of a failure in an interval.</td>
```
Is this a Mesos specific configuration? If we want to support that for all `ResourceManagers`, then it might be better to use a more generic configuration name. E.g. `resourcemanager.maximum-workers-failure-rate`.

Should we combine `maximum-failed-workers-per-interval` and `workers-failure-rate-interval` into a `failure-rate` config option which defines how many workers can fail per minute or per hour?

HuangZhenQiu replied:
Rather than separate configs for Yarn and Mesos, I added the failure rate option into resource manager config.

tillrohrmann commented on docs/_includes/generated/yarn_config_configuration.html:
```diff
@@ -27,6 +27,11 @@
+            <td><h5>yarn.containers-failure-rate-interval</h5></td>
+            <td style="word-wrap: break-word;">300</td>
+            <td>The interval for measuring failure rate of containers</td>
+        </tr>
```
Duplicate config option.

tillrohrmann commented on docs/_includes/generated/yarn_config_configuration.html:
```diff
@@ -42,6 +47,11 @@
+            <td><h5>yarn.maximum-failed-containers-per-interval</h5></td>
+            <td style="word-wrap: break-word;">-1</td>
+            <td>Maximum number of containers the system is going to reallocate in case of a failure in an interval.</td>
+        </tr>
```
Duplicate config option.

tillrohrmann commented on flink-mesos/src/main/java/org/apache/flink/mesos/configuration/MesosOptions.java:
```diff
@@ -99,6 +99,25 @@
+		key("mesos.workers-failure-rate-interval")
+			.defaultValue(300)
+			.withDeprecatedKeys("The interval for measuring failure rate of workers");
+
```
I would suggest to use a single `ConfigOption WORKERS_FAILURE_RATE` because it would make the configuration easier.

tillrohrmann commented on flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/MesosResourceManager.java:
```diff
@@ -663,7 +669,17 @@ public void taskTerminated(TaskMonitor.TaskTerminated message) {
+						failureInterval.toString()))));
+			} else {
+				startNewWorker(launched.profile());
+			}
```
This block seems to be generic enough that it should live in the base class.

tillrohrmann commented on flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/MesosResourceManager.java:
```diff
@@ -167,7 +169,9 @@ public MesosResourceManager(
-			JobManagerMetricGroup jobManagerMetricGroup) {
+			JobManagerMetricGroup jobManagerMetricGroup,
+			Time failureInterval,
+			int maxFailurePerInterval) {
```
either let's have `int failuresPerHour` or we should create a `FailureRate` class which encapsulates the `failureInterval` and `maxFailurePerInterval` values.

tillrohrmann commented on flink-mesos/src/test/java/org/apache/flink/mesos/runtime/clusterframework/MesosResourceManagerTest.java:
```diff
@@ -708,6 +720,50 @@ public void testWorkerFailed() throws Exception {
+			// wait for the registerSlotRequest completion
+			registerSlotRequestFuture.get();
+			assertEquals(0, rmServices.slotManager.getNumberPendingSlotRequest());
+		}};
```
I think it would be good to move this test into `ResourceManagerTest` because the functionality is not Mesos specific.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java:
```diff
@@ -145,6 +148,15 @@
 
+	protected final Time failureInterval;
+
+	protected final int maximumFailureTaskExecutorPerInternal;
```
typo

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java:
```diff
@@ -626,6 +686,44 @@ public void unRegisterInfoMessageListener(final String address) {
+		slotManager.rejectAllPendingSlotRequests(e);
+	}
+
+	protected synchronized void recordFailure() {
```
Let's call it `recordWorkerFailure()` and add some JavaDoc when to call it.

tillrohrmann replied:
Does it need to be synchronized?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java:
```diff
@@ -421,6 +474,13 @@ public void disconnectJobManager(final JobID jobId, final Exception cause) {
 					slotRequest.getJobId(),
 					slotRequest.getAllocationId());
 
+				if (shouldRejectRequests()) {
```
I think this check should rather be made in the `ResourceActionsImpl#allocateResource` because the slot request might still be fulfilled with an existing slot.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java:
```diff
@@ -178,6 +178,10 @@ public int getNumberPendingTaskManagerSlots() {
 		return pendingSlots.size();
 	}
 
+	public int getNumberPendingSlotRequest() {
```
Does this need to be public? We should add `@VisibleForTesting`.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java:
```diff
@@ -300,6 +304,18 @@ public boolean registerSlotRequest(SlotRequest slotRequest) throws SlotManagerEx
+		}
+
+		pendingSlotRequests.clear();
+	}
```
We should add a test for this method.

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java:
```diff
@@ -395,12 +405,20 @@ public void onContainersAllocated(List<Container> containers) {
+							rejectAllPendingSlotRequests(new MaximumFailedTaskManagerExceedingException(
+								String.format("Maximum number of failed container %d in interval %s"
+										+ "is detected in Resource Manager", maximumFailureTaskExecutorPerInternal,
+									failureInterval.toString()), t));
```
This branch should go into the `recordFailure` method.

HuangZhenQiu replied:
done

tillrohrmann commented on flink-yarn/src/test/java/org/apache/flink/yarn/YarnResourceManagerTest.java:
```diff
@@ -521,4 +533,56 @@ public void testOnContainerCompleted() throws Exception {
+	 * 	contains is hit.
+	 */
+	@Test
+	public void testOnContainersAllocatedWithFailure() throws Exception {
```
With the generalized test for the failure rate behaviour in `ResourceManagerTest` we would only need to check that a failing container would call `ResourceManager#recordFailure`.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java:
```diff
@@ -145,6 +148,15 @@
+
+	private boolean checkFailureRate;
+
+	private final ArrayDeque<Long> taskExecutorFailureTimestamps;
```
I think we should encapsulate the metering logic behind some interface. Ideally I would like to use exponentially-weighted moving average to approximate the failure rate. This would have the benefit that we would not have to store every timestamp here. However, with an interface we could also have different implementations (e.g. the exact one you have implemented here).

HuangZhenQiu commented on flink-yarn/src/main/java/org/apache/flink/yarn/configuration/YarnConfigOptions.java:
```diff
@@ -83,8 +83,27 @@
+	public static final ConfigOption<Integer> MAX_FAILED_CONTAINERS_PER_INTERVAL =
+		key("yarn.maximum-failed-containers-per-interval")
+		.defaultValue(-1)
+		.withDescription("Maximum number of containers the system is going to reallocate in case of a failure in an interval.");
```
Done

HuangZhenQiu commented:
@tillrohrmann @suez1224 
Thanks for reviewing the PR. According to your suggestions. I changed as
1) As the feature is generic for both Yarn and Mesos, add only the maximum failure rate config option in resource manager config. 
2) Put failure rate related logic into the TimestampBasedFailureRater which implements the FailureRater interface. As I don't want to mix two changes with different purpose in the same PR, we can make other code FailureRateRestartStrategy use it in another small PR.
3) For the failure rate test for RM, I tried to do in ResourceManagerTest. I found it is hard to mimic the behavior of registerSlotRequest without mocking lots components. And I also have to setup Test RM exactly like what YarnResourceManagerTest is doing. Thus, I still put the test cases separately in YarnResourceManagerTest and MesosResourceManagerTest.

HuangZhenQiu commented:
@tillrohrmann 
I tested the PR end to end in production. It works perfectly to handle hdfs unaccessible caused tm start failure. Please see the job log: https://www.dropbox.com/s/o169th1tudgw2q2/jobmanager.log?dl=0

I also merged with latest master. Please review it one more time. Thanks.

tillrohrmann commented:
Thanks for the update @HuangZhenQiu. I'll give it another pass now.

tillrohrmann commented on flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/MesosResourceManager.java:
```diff
@@ -663,7 +666,10 @@ public void taskTerminated(TaskMonitor.TaskTerminated message) {
+
+			if (recordFailure()) {
+				startNewWorker(launched.profile());
+			}
```
I think we should move the logic whether we can start a new worker or not to the `ResourceManager`. E.g. we could have the following methods in `ResourceManager`
```
protected Collection<ResourceProfile> tryStartNewWorker(ResourceProfile resourceProfile) {
	if (failureRater.exceedMaximumFailureRate()) {
		return Collections.emptyList();
	}

	return startNewWorker(resourceProfile);
}

protected void startNewWorkerIfNeeded(ResourceProfile resourceProfile, int pendingSlots) {
	int currentPendingSlots = pendingSlots;
	while (currentPendingSlots < getNumberRequiredTaskManagerSlots()) {
		final Collection<ResourceProfile> slots = tryStartNewWorker(resourceProfile);

		if (slots.isEmpty()) {
			break;
		}

		currentPendingSlots += slots.size();
	}
}
```

and then we only call `startNewWorkerIfNeeded` from here.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java:
```diff
@@ -425,6 +462,13 @@ public void disconnectJobManager(final JobID jobId, final Exception cause) {
+						new RuntimeException(String.format("Maximum number of failed container %d in interval %s "
+							+ " is detected in Resource Manager.", failureRater.getCurrentFailureRate(),
+							failureRater.getFailureInterval().toString()))));
+				}
```
I think this is the wrong place to check this. We should check the `failureRater.exceedMaximumFailureRater()` in a method from which we call `startNewWorker`. E.g. `tryStartNewWorker`.

```
protected Collection<ResourceProfile> tryStartNewWorker(ResourceProfile resourceProfile) {
	if (failureRater.exceedMaximumFailureRate()) {
		return Collections.emptyList();
	}

	return startNewWorker(resourceProfile);
}

protected void startNewWorkerIfNeeded(ResourceProfile resourceProfile, int pendingSlots) {
	int currentPendingSlots = pendingSlots;
	while (currentPendingSlots < getNumberRequiredTaskManagerSlots()) {
		final Collection<ResourceProfile> slots = tryStartNewWorker(resourceProfile);

		if (slots.isEmpty()) {
			break;
		}

		currentPendingSlots += slots.size();
	}
}
```

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java:
```diff
@@ -630,6 +674,29 @@ public void unRegisterInfoMessageListener(final String address) {
+	 * @return whether should acquire new container/worker after the failure
+	 */
+	@VisibleForTesting
+	protected boolean recordFailure() {
```
I think the boolean return type is not needed.

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java:
```diff
@@ -395,12 +402,14 @@ public void onContainersAllocated(List<Container> containers) {
+						if (recordFailure()) {
+							// and ask for a new one
+							requestYarnContainerIfRequired();
+						}
```
I think you are missing the `onContainersCompleted` case. If a container fails after the `startContainer` request has been sent, we won't record the failure and always allocate a new container. That's the reason why I would like to move this logic into the `ResourceManager` so that all container start call will go through the same code path.

tillrohrmann replied:
I think the changes are not properly tested. Otherwise `onContainersCompleted` should have been caught. Please add the relevant test cases.

tillrohrmann commented on flink-yarn/src/test/java/org/apache/flink/yarn/YarnResourceManagerTest.java:
```diff
@@ -518,4 +528,61 @@ public void testOnContainerCompleted() throws Exception {
+				rejectAllPendingRequestFuture.get();
+				assertEquals(rmServices.slotManager.getNumberPendingSlotRequest(), 0);
+			});
+		}};
```
This test is super hard to maintain with all this mocking. I would suggest to add a test for the `ResourceManager` which makes sure that you cannot start a new worker if the `FailureRater` exceeds the maximum failure rate. Additionally, we should add test cases for the `YarnResourceManager` which makes sure that `onContainersCompleted` and failures in the `onContainerAllocated` method will increase the `failureRater`.

tillrohrmann commented on flink-mesos/src/test/java/org/apache/flink/mesos/runtime/clusterframework/MesosResourceManagerTest.java:
```diff
@@ -708,6 +719,47 @@ public void testWorkerFailed() throws Exception {
+			// wait for the registerSlotRequest completion
+			registerSlotRequestFuture.get();
+			assertEquals(0, rmServices.slotManager.getNumberPendingSlotRequest());
+		}};
```
I think we can improve this test by only testing that `resourceManager.taskTerminated` increases the `failureRater` and then having a dedicated test for the `ResourceManager` which makes sure that you won't start new workers if `failureRater` is exceeded. Then we would not have to mock all this behaviour which is super brittle.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/failurerate/FailureRater.java:
```diff
@@ -0,0 +1,63 @@
+/**
+ * A rater to record the failure rate within a time interval.
+ */
+public interface FailureRater {
```
Maybe the `FailureRater` could still be useful if we use it to wrap a `Meter` and the maximum failure rate. Then the interface would offer
```
public interface FailureRater {
    boolean exceedsFailureRate();
    void markFailure();
}
```

tillrohrmann commented:
@HuangZhenQiu are you still working on this PR? I had some open comments which we should address before we can move forward.

HuangZhenQiu commented:
@tillrohrmann 
Yes, I am working on it. Will ping you back in weekend.

tillrohrmann commented:
Hi @HuangZhenQiu, could you make any progress?

HuangZhenQiu commented:
@tillrohrmann 
https://issues.apache.org/jira/browse/FLINK-12342 and https://issues.apache.org/jira/browse/FLINK-12343 last two weeks for internal users. Will continue to work on this this week.

tillrohrmann commented:
Great @HuangZhenQiu. Looking forward to your update :-)

HuangZhenQiu commented on flink-core/src/main/java/org/apache/flink/configuration/ResourceManagerOptions.java:
```diff
@@ -55,6 +55,18 @@
 			" Its not possible to use this configuration key to define port ranges.");
 
+	/**
+	 * Defines the maximum number of workers (YARN / Mesos) failure can happen in a minute.
```
From our production experience, it would be better to let the job failure immediately to make engineer to involve into the issue. The failure rate here is to just distinguish whether it is a transient failure or real issue caused by external systems such as hdfs. Thus, I am not sure it is needed for wait for failure rate falls below the  maximum.

HuangZhenQiu commented:
@tillrohrmann 
As this branch haven't been synced with Master for a while, I will create another Pull Request.

tillrohrmann commented:
You could also force push into this branch if you want. Do as you like @HuangZhenQiu and keep me updated.

HuangZhenQiu commented:
@tillrohrmann I do like your idea to put the failure rate check logic in RM, so that we can just add test case on ResourceManagerTest. It make the test cases maintenance much more easier. I changed the code accordingly. Please review it again, when you have time.

HuangZhenQiu commented on flink-runtime/src/main/java/org/apache/flink/runtime/failurerate/TimestampBasedFailureRater.java:
```diff
@@ -0,0 +1,82 @@
+	private final Time failureInterval;
+	private final ArrayDeque<Long> failureTimestamps;
+
+	public TimestampBasedFailureRater(int maximumFailureRate, Time failureInterval) {
```
The number helps to define a failureTimestamps for more accurate rate calculation than meter. It is also more the rater self-container for check exceedsFailureRate.

HuangZhenQiu commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java:
```diff
@@ -300,6 +305,25 @@ public boolean registerSlotRequest(SlotRequest slotRequest) throws SlotManagerEx
+			rejectPendingSlotRequest(pendingSlotRequest, cause);
+
+			// notify each job master about this exception
+			resourceActions.notifyAllocationFailure(
```
If you see other part of the class. resourceActions.notifyAllocationFailure are called separately. How about keep the same pattern here?

HuangZhenQiu commented:
@tillrohrmann 
After the rebase master, the conflict is mistakenly reported. I created another PR for FLINK-10868.
https://github.com/apache/flink/pull/8952
